Add show() method as preferred entrypoint

duytnguyendtn opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

## Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

### Backstory
Ok I'm going to finally admit to a nearly 18 month secret I've been hiding since I originally wrote the first helper for specviz. I've always had negative opinions about `viz.app`. I feel like it's not intuitive and not apparent to a brand new user, and almost forces them to read the documentation first before even trying to take it for a spin (which I imagine some may see as a good thing, but I'm going to argue I don't have to read Chrome's documentation when I launch it!). So I added `specviz.show()` as a wrapper to the `specviz.app` call:
https://github.com/spacetelescope/jdaviz/blob/392e1b626c2ac0dbe314bb103cadf001e1b9db4c/jdaviz/configs/specviz/helper.py#L200-L201

Here's the secret... it doesn't work. It never worked. I struggled for SO LONG at the time but I couldn't fix it; I just didn't understand what front-end wizardry made `viz.app` work.. So... I kept my mouth shut because I felt so strongly it should be there, and swore one day I would fix it. You can even see how old this is because it's only in the *Specviz* helper, and not in the base helper class; this was before I even wrote the base helper class, and the only helper we had was Specviz!

With @eteq's introduction of #952, I now further think this should be the way to show our viz tools, because a helper method would allow us to add options and arguments to a SINGLE entry point, rather than separate helpers for each launching mode (i.e. `viz.show(mode=sidecar)`, `viz.show(new_tab=True)` rather than separate `viz.show_in_new_tab()` and `viz.show_in_sidecar()`

In my review of #952, I had a real *bruhhh* moment. The reason why it never worked was because I never **returned** the app...

... They say the mistakes of the past show how much you've grown. As an optimist, I choose to view it through this lens 😅 I'm legitimately giddy seeing this PR finally in fruition. Now I can finally put this issue to rest...

As a warning, I WILL die on top of this hill!! BYOP (Bring Your Own Pitchforks!)

### tl;dr What actually changed??
This PR adds `viz.show()` as a helper method to the base ConfigHelper class, and updates all notebooks and documentation to show `viz.show()` rather than `viz.app`

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

## Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

PatrickOgle commented:
I have no objection, as long as viz.app still works, for back-compatibility in published notebooks.

kecnry commented:
I have two hesitancies, which don’t necessarily mean you have to leave your hill:
1. Even though viz.app is still there, this hides that a bit which might make some advanced functionality a little more hidden (or maybe that’s a good thing since users will need to know what they’re doing before digging deeper)
viz.show() doesn’t technically show anything, it just returns the app and the notebook interprets that as something to show.  
2. If you were to run this outside a notebook, it would just return an object.  Is there anyway to have this either raise a warning/error in python or (better yet) launch a popup window in a kernel?

duytnguyendtn commented:
@PatrickOgle: Yes! This entirely maintains backwards compatibility. This is just an additional helper method, and as with all helper methods, it's just for convenience
@kecnry Here's my thoughts to your points:
1. I feel that the helper's goal **is** to hide the base level `viz.app`! The whole point of the convenience helpers is to provide simple entrypoints to jdaviz functionality without having to drill down. Once you have to drill down, I'm going to say it's a feature that they need to read the docs for more advanced functionality
2. Actually, my aha! moment was assuming interpretation of the application class was exactly how `viz.app` worked! When I save a rendered notebook and try to open it up, both methods briefly show the returned Application object, before the app is actually rendered. I assumed this was exactly what you were saying, as the notebook interpreting the object as something to show. But if I'm wrong, please correct me! Oh frontend wizard, teach me your secrets!
![image](https://user-images.githubusercontent.com/25206008/141311932-3f2b503e-627a-49e0-881b-4b1c39e84fba.png)
3. Clarifying question about your last point: When outside the notebook context, the behavior of `viz.show()` and `viz.app` is equivalent here; both return the application itself. Are you suggesting that this shouldn't be the case for *either* and an exception should be thrown in both scenarios, or that this behavior is OK for `viz.app` and not `viz.show()`?
```
>>> viz.app
Application(components={'g-viewer-tab': '<template>\n  <component :is="stack.container">\n    <g-viewer-tab\n ...

>>> viz.show()
Application(components={'g-viewer-tab': '<template>\n  <component :is="stack.container">\n    <g-viewer-tab\n ...
```

kecnry commented:
@duytnguyendtn - Exactly, the notebook knows to interpret that returned object and display it.  I guess my slight issue is that using the "verb" `show` suggests that jdaviz is doing the showing.  From the user-perspective of course, this distinction probably doesn't matter: in the notebook the end result is the UI is _shown_ and no one cares how it happens.  But, if you run the same lines in python (outside of a notebook), nothing shows which might be more confusing now because we hide this distinction behind the verb.  Or if you do `blah = viz.show()` in a notebook, nothing is shown but rather is equivalent to `blah = viz.app`.  Compare this to matplotlib's show, which will popup a standalone window from python and shows in a notebook regardless if you assign the output, for example.  I don't think it's necessarily a dealbreaker, and think the pros likely outweigh the cons, I just wanted to bring that up for discussion/consideration.

duytnguyendtn commented:
For an outside interpreter, I wonder if I can trigger the desktop app to fire with the user's instance of the viz tool. I feel that would adequately answer that usecase. It looks like we might want to add some "environment detection", i.e.:
```
if environment == notebook: rely on interpreter
elif environment == python (but outside notebook): start desktop app
...
```
I'm not sure what to do with `blah = viz.show()`... That one's tricky.

camipacifici commented:
No objections. I do not see how our current users (even the advanced ones) would complain about this.
I would like to better understand what magic you can now do with show(), like can you actually feed a file to show() so that the app opens and the data are loaded with one command?

kecnry commented:
@duytnguyendtn I like that idea, except that the desktop app would need to communicate with the original instance which was likely started without a kernel, so that might be quite difficult to implement in the current infrastructure.  For now maybe its worth just detecting if the current session does not support showing and raise an error during `viz.show()` that points to `viz.app` to access the underlying object or suggests running in Jupyter?  I don't think there's much we can do about the `blah = viz.show()` scenario unless we instead have `show` actually inject into the cell or something instead of just returning, but I also don't think its a showstopper and the advantages are probably well worth it.

duytnguyendtn commented:
@kecnry I opted to raise a warning instead and return the app anyways just in case the user has an interactive environment that we don't know about. What do you think?

@camipacifici Ooh, I like the idea of `viz.show(data="myfile.fits")` as a oneliner to show both the data AND the viz tool! I don't think I'll implement that here, since that might be a larger discussion. I think the best manifestation of the improvements this PR brings (beyond the naming) can best be seen in #952. Erik has two new api methods to display the app in different ways (`viz.show_in_new_tab()` and `viz.show_in_sidecar()`). Notice how these are two separate methods. By having a central `viz.show()` we can avoid this, and rather just change the arguments (i.e. something like `viz.show(mode=sidecar)` or `viz.show(mode=new_tab)`. But really, my main horse in this race is the naming clarity (i.e. `viz.show()` is more clear to the user, and also matches other python programs, like `matplotlib`)

ibusko reviewed (dismissed):
Code looks good to me, and works.

pllim commented:
Need a rebase to resolve conflicts.

mariobuikhuizen commented:
@duytnguyendtn nice to see you got a secret off your chest 😄

You could also use `display()` to show the app and not return the app object (this is what implicitly happens with the last line of a notebook cell). This would also allow for `.show()` not having to be the last code line in a cell.

```python
  def show(self):
    display(self.app)
```

pllim commented on CHANGES.rst:
```diff
@@ -32,6 +32,8 @@ API Changes
 - Viewers now can access the calling Jdaviz application using
   ``viewer.session.jdaviz_app``. [#1051]
 
+- Viz tool display changed to ``viz.show()``. [#965]
```
I think this should be merged with the change log for #952 . Also you need to rebase against latest `main` branch upstream (not your fork). I see the change log here has not picked up recent fixes.

pllim replied:
Yeah, I agree with Duy.

duytnguyendtn commented:
Moved to draft to address discussion for display modes referenced in https://github.com/spacetelescope/jdaviz/pull/952#issuecomment-1024506439

pllim commented on jdaviz/tests/test_show.py:
```diff
@@ -1,17 +1,32 @@
+
+
+def test_show_new_browser_tab(specviz_helper):
+    with warns(RuntimeWarning):
```
If you want to be more explicit, you can also try to match the warning message as follows. It also takes regex if you wish.
```suggestion
    with warns(RuntimeWarning, match='Error detected in display'):
```

duytnguyendtn commented:
I put this as draft due to getting feedback on the API, but I'm afraid this might have just dropped this PR off everyone's radar. I'm moving this back out of draft to solicit feedback on the API. Thanks! (pinging @eteq one more time for the review :) )

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -274,96 +275,88 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
         -------
-        sidecar
-            The ``sidecar.Sidecar`` object used to create the tab.
+        sidecar or Nothing
```
```suggestion
        sidecar or None
```

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -274,96 +275,88 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
-        else:
-            kwargs['anchor'] = 'tab-after'
-        return self.show_in_sidecar(**kwargs)
+        except Exception as e:
```
might be nice to catch particular exceptions here (since the try/except wraps a lot more code than just the lines that do the displaying).

duytnguyendtn replied:
So I left it generic because I effectively wanted any errors to be thrown back to the user, but to always fallback to the IPython display. I figured if anything errors out, the user can be responsible for trying to fix it. Unless there are specific cases you're thinking that should be caught here?

kecnry replied:
I guess right now there isn't much else that could raise an error in that block.  My main concern would be any error introduced by us to the code in the future, or a ctrl+c type exception (although that seems quite unlikely since nothing should ever hang or take significant time).  My second concern is if warnings are hidden/ignored, then it might be difficult for the user to know why it's not behaving as expected - especially in the "unrecognized display mode" case where a simple typo could trigger that.  I'm curious what others think though.

kecnry replied:
Yeah, I suspect I'm in the minority on preferring errors to warnings.  So I won't insist on changes if this was by design.

duytnguyendtn replied:
Ironically, I think I'm actually on your side here. I originally caught the error, fell back on the IPython display, THEN reraised the exception to the user to be more explicit. It was just... distasteful in how it appeared. Pey Lian and I had a LONG thread on this last Friday... If you can find a way to better represent it in a way I coudn't, please do!

kecnry replied:
ok, then let's leave it as-is if no one else has objections/suggestions.

pllim replied:
To be clear, and perhaps I wasn't being clear in my original communication, my original suggestion was to just error out with the exception (i.e., no try-except). However, Duy wanted to fall back to IPython, so in that case I suggested turning exception into warning and then do the fallback. It just does not make sense to me to throw and exception and then fallback to IPython. Perhaps I have missed something here.

kecnry reviewed (commented):
Definitely like the way the API is converging to handle all these options under one "roof"!  Just a few small comments and then I'll give it a spin to test once its rebased.

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -274,96 +275,88 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
+            elif mode == "new jupyter tab":
+                if 'anchor' in kwargs:
+                    if 'tab' not in kwargs['anchor']:
+                        raise ValueError('new jupyter lab cannot have a non-tab anchor')
```
```suggestion
                        raise ValueError('new jupyter tab cannot have a non-tab anchor')
```

eteq reviewed (dismissed):
Overall I like this! But several changes to suggest around the edges.

In addition to some of my fiddly naming change suggestions inline, I have a more general suggestion about the sidecar modes: in addition to just allowing `sidecar`, also allow `sidecar-BLAH` where `BLAH` is one of the options split-right', 'split-left', 'split-top', 'split-bottom', 'tab-before', 'tab-after', 'right' (and disallow `anchor` as a keyword argument).  That's trivial to do because it's just the `anchor` options, and it combines everything cleanly.

You should also make sure those options are listed explicitly - as @pllim pointed out in the #952 review it's hard to find those docs without digging, so we should just state it since it's not that likely to change. (I mentioned this inline too)

One other top-level thought: it *might* be desirable to follow a separation-of-concerns convention of having `show` just call various private methods like `_show_inline` / `_show_sidecar`/etc.  Then there's less if/elif/else logic, and then `show` can be a one-liner like ``return getattr(self, '_show_' + mode)(*args, **kwargs)``.  That's sort of an internal detail but might make it a little bit easier to add more

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -274,96 +275,88 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
 
         Parameters
         ----------
+        mode : str
```
Or even `location` (`loc`)?

duytnguyendtn replied:
I was thinking the same thing @kecnry. My brain immediately jumped to `location`. Ok with you @eteq ?

duytnguyendtn replied:
Implemented in https://github.com/spacetelescope/jdaviz/pull/965/commits/866a3fc0204916977082838b1e7fcd925abaad36

duytnguyendtn commented on jdaviz/core/helpers.py:
```diff
@@ -274,96 +275,88 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
+                `jupyterlab-sidecar <https://github.com/jupyter-widgets/jupyterlab-sidecar>`_
+                for the most up-to-date options.
+
+            "new jupyter tab": Display the Jdaviz application in a new tab in JupyterLab.
```
My intention was to separate between a new `sidecar` tab within Jupyter Lab vs a new `browser` tab, which I see being maybe possible with the work @mariobuikhuizen and @maartenbreddels

duytnguyendtn replied:
Largely no longer relevant after I switched to the `sidecar:<anchor_here>` convention

duytnguyendtn commented on jdaviz/core/helpers.py:
```diff
@@ -274,96 +275,88 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
+            "sidecar": Display the Jdaviz application in a "sidecar", which
+            by default is a tab on the right side of the JupyterLab  interface.
+
+                Additional keywords not listed here are passed into the
```
Added!

duytnguyendtn commented:
This PR is ready to review again!

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,96 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
+                raise NotImplementedError
 
-        Returns
-        -------
-        sidecar
-            The ``sidecar.Sidecar`` object used to create the tab.
+            elif loc == "popout":
+                raise NotImplementedError
```
to be analogous to the sidecar anchors, I'd suggest something like "popout:window" and "popout:tab"... but since they're not implemented or even documented here, that could be easily changed later (I just at least wanted to leave a papertrail while its still fresh).

pllim replied:
I don't understand why these need special `elif`. Why not just let them go through the same logic as other invalid inputs, given they are not even documented?

duytnguyendtn replied:
Agreed! I think if we end up implementing different popout locations, I'd be a big fan of the above. I mainly put this codeblock here just to remind us to add this later

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,96 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
-    def show_in_new_tab(self, **kwargs):
-        """
-        Display the Jdaviz application in a new tab in JupyterLab.
+                scar = Sidecar(**kwargs)
```
If `Sidecar` only takes `anchor` and `title`, then it might make sense to just pass them directly here and have a kwarg in the method that defaults to `title=None`.  That way typos in arguments are caught and not just absorbed by the catch-all `**kwargs`.

duytnguyendtn replied:
Just checked in `sidecar` and I could only find `anchor` and `title` so let me follow your suggestion here ^

duytnguyendtn replied:
Done!

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,96 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
-            if 'tab' not in kwargs['anchor']:
-                raise ValueError('show_in_new_tab cannot have a non-tab anchor')
-        else:
-            kwargs['anchor'] = 'tab-after'
-        return self.show_in_sidecar(**kwargs)
+        Preserved for backwards compatibility
+        Shows Jdaviz in a sidecar with the default anchor: right
+        """
+        return self.show(loc="sidecar", **kwargs)
+
+    def show_in_new_tab(self, **kwargs):
+        """
+        Preserved for backwards compatibility
+        Shows Jdaviz in a sidecar in a new tab to the right
+        """
+        return self.show(loc="sidecar:tab-after", **kwargs)
```
should we deprecate these somehow so we can remove them eventually?  Were they ever actually used in any documentation/notebooks?

pllim replied:
Yes, Erik added them into example notebooks but I am not sure who actually uses it.

rosteen replied:
And they've annoyed me by being there ever since! 😆

pllim replied:
Me too!

duytnguyendtn replied:
Seems like there's a desire to remove at least the notebook entries to start deprecating them. I, for one, am happy to oblige!

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,96 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
-    show_in_sidecar.__doc__ = show_in_sidecar.__doc__.format(
-        sidecar_anchor_values=repr(Sidecar.anchor.values)[1:-1])
+                # If title unset, default to the viz config
+                if 'title' not in kwargs:
+                    kwargs['title'] = self.app.config
```
This should be a little more efficient, but effectively does the same thing:

```suggestion
                kwargs.setdefault('title', self.app.config)
```

(but might be irrelevant depending on what you decide on my comment about `**kwargs`)

duytnguyendtn replied:
Dropping **kwargs, so this is irrelevant as pointed out above

kecnry reviewed (approved):
So much cleaner and convenient!  As far as I can tell, this works great - I just have a few minor comments which you might want to consider before merging, but you can take them or leave them.

I still would personally prefer getting rid of the try/excepts entirely and letting errors be raised.  Using any of the sidecar location in notebook already know to fallback on inline display without the try/excepts (and currently with the try/excepts don't even raise a warning because of that).

pllim commented on jdaviz/tests/test_show.py:
```diff
@@ -1,17 +1,27 @@
 import sidecar
+from pytest import warns
```
This is easier to read:

```python
import pytest

with pytest.warns(...):
    test_code_here
```

duytnguyendtn replied:
Done!

pllim commented on notebooks/CubevizExample.ipynb:
```diff
@@ -54,7 +50,7 @@
    "outputs": [],
    "source": [
-    "viz.show_in_sidecar()"
+    "viz.show(loc='sidecar')"
```
I really don't like these being run-able cells in example notebooks. They are already documented in `display.rst`. If you must keep them, please change them into "raw nbconvert" cell type, so people don't accidentally run all the ways to display because they are not supposed to.

duytnguyendtn replied:
Removed these cells in all the notebooks!

pllim commented on notebooks/CubevizExample.ipynb:
```diff
@@ -63,7 +59,7 @@
    "outputs": [],
    "source": [
-    "viz.show_in_new_tab()"
+    "viz.show(loc='sidecar:tab-after')"
```
See above.

duytnguyendtn replied:
Removed these cells in all the notebooks!

pllim reviewed (commented):
This PR should use the *Squash and Merge* button when it is ready for merge.

Overall, I think it is great that you consolidated similar APIs into one. Thanks!

I will withhold second approval until Kyle's comments are addressed but I have dismissed Erik's very stale "changes requested". I agree with Kyle's comments about exception handling.

Could you please break the very long RST paragraphs into shorter lines? It will make future diff reviews easier.

I would vote to deprecate `show_in_sidecar` and `show_in_new_tab` as part of this PR, and remove them from example notebooks. You can use https://docs.astropy.org/en/stable/api/astropy.utils.decorators.deprecated.html for convenience if you want.

More comments below.

kecnry commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,96 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
         -------
-        sidecar
-            The ``sidecar.Sidecar`` object used to create the tab.
+        sidecar or None
```
I agree with this, the `with scar: display(self.app)` should be all that's needed.  The other modes don't return anything.

duytnguyendtn replied:
This was behavior I carried over from #952. Seems the desire is to return nothing. I can change that

pllim replied:
I seriously doubt anyone would directly interact with the sidecar object. My vote is not to return it.

duytnguyendtn commented on docs/notebook/display.rst:
```diff
@@ -0,0 +1,30 @@
+
+You can additionally specify the location with the ``loc`` argument. For example, ``inline`` can be specified manually with::
+
+    imviz.show(loc='inline')
```
Point raised https://github.com/spacetelescope/jdaviz/pull/965#discussion_r808364655 by @eteq, with suggestion by @kecnry. I think I've seen `location` abbreviated as `loc` before, hence why I jumped on it, though I generally prefer avoiding abbreviations, so if @pllim you think `location` would be better, I'd be more than happy to expand it!

pllim replied:
According to Python zen, it is better to be explicit, so "location" would be the clearest. But I won't block the merge if you prefer "loc". Either way, once we merge, the boat will sail; we should not go back and forth after merge.

tl;dr -- I'll leave the final name to your discretion.

duytnguyendtn commented:
Ok I *think* I got all the review comments. I'll watch the tests to make sure I didn't miss anything, but should be good for a second round of reviews!

kecnry commented on jdaviz/tests/test_show.py:
```diff
@@ -1,17 +1,35 @@
+               'tab-before', 'tab-after', 'right']
+    for anchor in anchors:
+        res = specviz_helper.show(loc=':'.join(['sidecar', anchor]))
+        assert isinstance(res, anchor)
```
I don't think this will work... `anchor` is a string in the loop.

duytnguyendtn replied:
Am I being dumb?
```
Python 3.10.5 (tags/v3.10.5:f377153, Jun  6 2022, 16:14:13) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> ':'.join(['sidecar', "split-right"]) 
'sidecar:split-right'
>>>
```

pllim replied:
What does `isinstance(res, anchor)` mean?

kecnry replied:
Yes, I meant the `isinstance` line, the second argument to `isinstance` needs to be the type/class you're comparing it to, not a string.  Is there some way to access the anchor after the sidecar object is instantiated (something like `res.anchor == anchor`?  

As for the other line, that should work with the `.join`, but might be a little easier to read as `f"sidebar:{anchor}"`.

duytnguyendtn replied:
Oh whoops, totally misread that lol. Thanks 😅

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,94 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
-        `jupyterlab-sidecar <https://github.com/jupyter-widgets/jupyterlab-sidecar>`_
-        for the most up-to-date options.
+    def show(self, loc="inline", title=None):
+        """ Display the Jdaviz application.
```
```suggestion
        """Display the Jdaviz application.
```

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,94 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
         -------
-        sidecar
-            The ``sidecar.Sidecar`` object used to create the tab.
+        scar : sidecar or `None`
```
So you want to keep returning this?

duytnguyendtn replied:
Consensus earlier was to not. Working on that now!

duytnguyendtn replied:
Implemented in https://github.com/spacetelescope/jdaviz/pull/965/commits/117d2ab00181f7ac4244560fe9a9927672c7b930

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,94 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
-        If this method is called in the "classic" Jupyter notebook, the app will
-        appear inline, as only lab has a mechanism to have multiple tabs.
+            else:
+                raise ValueError("Unrecognized display location: " + str(loc))
```
```suggestion
                raise ValueError(f"Unrecognized display location: {loc}")
```

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,94 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
+                          RuntimeWarning)
+            display(self.app)
+
+    def show_in_sidecar(self, anchor=None, title=None):
```
Do we want to deprecate this?

duytnguyendtn replied:
See https://github.com/spacetelescope/jdaviz/pull/965#discussion_r913820210

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,94 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
-        show_in_sidecar
-        show_inline
+        except Exception as e:
+            warnings.warn(f'Error detected in display: {repr(e)}. Falling back to IPython display',
```
@kecnry , are you okay with keeping this or you still rather it throw an exception?

duytnguyendtn replied:
Implemented in https://github.com/spacetelescope/jdaviz/pull/965/commits/e43864ed25bb6a3728afcd038e04236b624d05c2 @kecnry

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,94 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
+        location = 'sidecar' if anchor is None else f"sidecar:{anchor}"
+        return self.show(loc=location, title=title)
+
+    def show_in_new_tab(self, title=None):
```
Do we want to deprecate this?

duytnguyendtn commented:
Ok, I think I addressed everything in https://github.com/spacetelescope/jdaviz/pull/965#pullrequestreview-1026483132 so long as @pllim is okay with the Deprecation Warning? Since we removed the sidecar return, I also had to remove a lot of the coverage, and I think we're still trying to figure out how to test UI elements so I think the coverage will have to suffer a little bit to get this in. Otherwise, we're good for the final review!

pllim commented on CHANGES.rst:
```diff
@@ -46,6 +46,9 @@ API Changes
   have been renamed to "Colormap" and "Monochromatic," respectively, for all image
   viewers. [#1406]
 
+- Viz tool display changed to ``viz.show()``. Sidecar no longer returned by 
```
I thought I suggested this before. Maybe lost in rebase?

```suggestion
- Viz tool display changed to ``viz.show()`` from ``viz.app``. Sidecar no longer returned by 
```

pllim commented on jdaviz/core/helpers.py:
```diff
@@ -273,96 +274,94 @@ def get_model_parameters(self, models=None, model_label=None, x=None, y=None):
+
+            NOTE: Only applicable to a "sidecar" display.
 
         Returns
```
GitHub won't let me enter as suggestion, but you have to delete the whole "Returns" section because it no longer applies.

duytnguyendtn replied:
Oops, thanks for catching that!

pllim commented on jdaviz/tests/test_show.py:
```diff
@@ -1,17 +1,11 @@
 
-def test_show_inline(specviz_helper):
-    res = specviz_helper.show_inline()
+def test_show_popout(specviz_helper):
```
Does this test actually exposes a bug here? I thought you wanted "popout" to emit a `NotImplementedError`.

pllim reviewed (commented):
I think we're very close. Just a few more comments. 😅

pllim reviewed (approved):
Let's get this in for the release. Thanks!

duytnguyendtn commented:
Hehe... whoops, you beat me to the punch! Hopefully https://github.com/spacetelescope/jdaviz/pull/965/commits/82c0fe407aadcdb82708c53579c2f08b8a9835a1 passes the tests!

So I was going to reply to your comment https://github.com/spacetelescope/jdaviz/pull/965#discussion_r914166740 that the reason I was trying to catch a RuntimeError is because I chained the exception to explicitly say that there was an error in the displaying; hence I wrap the original error with a generic RuntimeError with a more user friendly explanation. 

But that got me thinking the test *should* check for the chained NotImplementedError so I pushed 82c0fe407aadcdb82708c53579c2f08b8a9835a1 which changes that test to check for that chained NotImplementedError

Just adding context here in case we suddenly find the tests are breaking at release, I know where it might be 😅

duytnguyendtn commented:
But that got me thinking the test *should* check for the chained NotImplementedError so I pushed I do have https://github.com/duytnguyendtn/jdaviz/commit/aa658ea7df9fbaa477d6fc2391541602ed017b59 which changes that test to check for that chained NotImplementedError. Didn't quite make it to the merge. Did we want to merge that to resolve https://github.com/spacetelescope/jdaviz/pull/965#discussion_r914166740?
